src/Stage/Desugar.elm

```diff
@@ -180,22 +181,67 @@ desugarExpr modules thisModule located =
             return Canonical.Unit
 
         Frontend.Record bindings ->
-            bindings
-                |> List.map (Binding.map recurse >> Binding.combine)
-                |> Result.combine
-                |> map
-                    (\canonicalBindings ->
-                        canonicalBindings
-                            |> List.map (\canonicalBinding -> ( canonicalBinding.name, canonicalBinding ))
-                            |> Dict.fromList
-                            |> Canonical.Record
-                    )
+            case maybeDuplicateBindingsError bindings of
+                Just error ->
+                    Err error
+
+                Nothing ->
+                    bindings
+                        |> List.map (Binding.map recurse >> Binding.combine)
+                        |> Result.combine
+                        |> map
+                            (\canonicalBindings ->
+                                canonicalBindings
+                                    |> List.map (\canonicalBinding -> ( canonicalBinding.name, canonicalBinding ))
+                                    |> Dict.fromList
+                                    |> Canonical.Record
+                            )
 
 
 
 -- HELPERS
 
 
+{-| Ensure that there are no two bindings with the same name.
+
+NOTE: The function will produce an error only for the _first_ duplicate pair.
+Subsequent duplicate pairs are ignored.
+
+-}
+maybeDuplicateBindingsError : List (Binding Frontend.LocatedExpr) -> Maybe DesugarError
+maybeDuplicateBindingsError bindings =
+    bindings
+        |> findDuplicatesBy .name
+        |> Maybe.map
+            (\( first, second ) ->
+                DuplicateRecordField
+                    { name = first.name
+                    , firstOccurrence = Located.replaceWith () first.body
+                    , secondOccurrence = Located.replaceWith () second.body
+                    }
+            )
+
+
+{-| Find the first two elements in a list that duplicate a given property.
+
+Implemented to avoid allocations.
+
+-}
+findDuplicatesBy : (a -> comparable) -> List a -> Maybe ( a, a )
+findDuplicatesBy property list =
```

Comment:
O(n^2) complexity... I wonder if there's a better way. Not sure yet

Reply:
Maybe be first ordering the list by field name? It would be O(nlog(n)).

Reply:
https://ellie-app.com/7mJmZBnbQKCa1
:arrow_up: Possible alternative (I don't like it though) but I'm not sure it's better. Also I'm getting a bit lost in the complexity of Dict.get, filter, etc... It's O(n log n) too, right?
We should run a Benchmark on the two (three if we want to try the sorting) implementations!

Reply:
If performance is a problem, we might want to run a couple of benchmarks before thinking complexity.

Reply:
Dict.get should be O(log(n)) (with n the number of (key,value) couples)

I don't see how Dict.filter could be less than O(n).

Reply:
But yes as pointed out by @xarvh it is only a performance issue, not sure we should be really concerned by this right now since the records dont containt a huge number of fields most of the time, so the O(n^2) behavior should not be an issue (moreover in you Ellie @Janiczek you do a lot of map/filter/... copying the lists which can be worst than the O(n^2) if the n is not big).

That said, did you consider using List.Extra.unique* fonctions (the `allDifferent` function seems to be exactly what we need but the implementation isn't very efficient)?

https://github.com/elm-community/list-extra/blob/8.2.2/src/List/Extra.elm#L415-L456

Reply:
So, I got curious.

TL;DR: The O(n^2) `onn` algorithm is faster only up to 8 items. At 9 items, `sort` is faster.

I'm not sure if there is any reliable way to estimate memory footprint, my guesstimate is that `onn` won't allocate much and `sort` transforms the list into a js Array which is still O(n) and js will be optimized af so it's good.

I was hoping to get the best from both worlds and run one algorithm for long lists and the other for short lists, but estimating the list length is in itself expensive, so meh.


![image](https://user-images.githubusercontent.com/1945841/69912081-f16c3480-1424-11ea-8471-1d31a0ab4744.png)

https://gist.github.com/xarvh/2e3ddf2e3f0f2c6e1f84505e1bff1f1b

Reply:
@xarvh Thanks for doing the benchmark!  I am not surprised that `sort` does better than o(n*n) for large numbers, of course, but I was surprised the threshold was so low.  My vote would be to use `sort` here, even though I understand the argument that most records are pretty small (and, probably, in fact fewer than 8 fields 99% of the time).  The `sort` code should be pretty easy to read and understand, and I don't like surprises for the worst case scenario.

This is obviously suggestive and just my opinion.  I think the n-squared approach is also fine.

Reply:
Agreed. Sort it is. We can always switch algorithm later on. I'll link the benchmark gist in the comments.

Reply:
Thanks!

Reply:
Should be mergeable.

Reply:
Can we commit this benchmark somewhere?

Reply:
Sure.